common-ts/src/clients/marketDataFeed.ts

```diff
@@ -0,0 +1,710 @@
+
+To ensure that trades and candles are in sync, we want to make sure that subscriptions for these things (which is done seperately) actually use the same websocket connection. Which is why this class is necessary.
+
+## Explanation of terminology:
+- A "subscription" is an actual underlying websocket connection to the Drift Data API. A subscription may have multiple subscribers attached to it, where the data coming through that subscription is sufficient for the subsciption config. A Subscription can have the data for both CANDLES and TRADES subscribers.
+- A "subscriber" is a specific instance of an external party making a subscription. It is either a TRADE or CANDLE subscriber.
+- A "subscriber subscription" represents the subscription information that we return to an external subscriber. It contains the subscription ID they can use to unsubscriber, an an observable to listen to the data.
+
+## Explanation for tradeSubscriptionLookup and candleSubscriptionLookup:
+This class supports subscribing to multiple arbitrary markets (and candle resolutions) at the same time. It also supports sharing a single websocket connection between multiple subscribers where possible. This creates one bit of complexity where we need to search for existing "compatible" subscriptions whenever a new subscriber comes through. Example:
+- Imagine a CANDLE subscription for SOL-PERP-Resolution-1 (Subscription A).
+- When a TRADE subscription comes through for SOL-PERP (Subscription B) we want it to use the same websocket connection as Subscription A, because they are for the same market - even though they are different subscription types. We use the `tradeSubscriptionLookup` to see that there is an existing subscription for this market and then attach the new trade subscriber to it.
+- When another CANDLE subscription comes through for SOL-PERP-Resolution-5 (Subscription C), there is no existing subscription in the `candleSubscriptionsLookup`, because the resolution is different. Hence we create a new subscription.
+- TO SUMMARISE :: A trade subscription can share any subscription with the same market. But a candle subscription must share the same market AND resolution. We use seperate lookup tables to index into compatible subscriptions.
+
+There is a minor inefficiency remaining in this implementation, where to be optimal we may want to "re-organise" subscriptions when they close. E.g.:
+- <open> SOL-PERP-Resolution-1 (Subscriber A) => will create new Subscription A
+- <open> SOL-PERP-Resolution-5 (Subscriber B) => will create new Subscription B
+- <open> SOL-PERP-TRADE (Subscriber C) => will use existing Subscription A
+- <close> SOL-PERP-Resolution-1 (Subscriber A) => will keep Subscription A open because Subscriber C is still listening
+- * We could more optimally close Subscription A and make Subscriber C use Subscription B
+- * We're not doing this for the same of simplicity. In the future we could add this optimisation if it becomes necessary.
+*/
+
```

Comment:
while this overall documentation helps a lot, I'd really wished that individual classes and methods are documented too